content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-certs.md

```diff
@@ -275,3 +281,23 @@ Such a controller is not a secure mechanism unless it not only verifies the Comm
 in the CSR but also verifies the requested IPs and domain names. This would prevent
 a malicious actor that has access to a kubelet client certificate to create
 CSRs requesting serving certificates for any IP or domain name.
+
+## Troubleshooting kubelet client certificates rotation {#kubelet-client-cert}
+
+By default, kubeadm configures a kubelet with automatic rotation of the client certificate located
+in `/var/lib/kubelet/pki/` and path specified in `/etc/kubernetes/kubelet.conf`. If this rotation process
+fails you might see errors such as `x509: certificate has expired or is not yet valid` in kube-apserver logs.
+To fix the issue, first check that `/var/lib/kubelet/pki/kubelet-client-current.pem`
+is a symlink. If not, backup and delete it. The kubelet will recreate it.
+Then follow these steps:
+1. Backup and delete `/etc/kubernetes/kubelet.conf` and regenerate it with
+`kubeadm init phase kubeconfig kubelet`. Pass the original `--config` file you used with `kubeadm init`. If you don't have it, manually edit the `kubelet.conf`.
+2. Restart the kubelet (`systemctl restart kubelet`) and wait for
+`kubelet-client-current.pem` to be recreated.
+3. Run `kubeadm init phase kubelet-finalize all` using `--config` file if necessary.
+4. Restart the kubelet again
+
+Depending on the root cause of the failure, it might be necessary to
+`kubeadm certs renew all` after the kubelet certificate is fixed, and then restart
+kubelet and kube-system PODs (or reboot the node).
+Note that on high availability setups, this must be done on all nodes.
```

Comment:
i have tested recovery of expired kubelet certificates before. also we have recommended the above steps to other users, prior to your report. the `kubeadm certs renew all` and static pod restarts were not needed.

are you sure something else was not at play?

Reply:
I'm 100% sure `kubeadm cert renew all` before the kubelet cert renew steps did not nothing for me, and after kubelet cert renew I restarted everything and it was still not working, and after I kubeadm cert renew again it worked. But as you said, I'm not sure if there were other factors. My kubelet user changed, when I recreated a kubelet.conf, don't know if it's related.

On the one hand my suggestion is probably not very sound, on the other hand it worked for me and it is a troubleshooting section after all. You tell me :shrug:

Reply:
> My kubelet user changed, when I recreated a kubelet.conf, don't know if it's related.

nope, i don't think the apiserver would have to be restarted due to that. it would accept the new client connection as long as the credentials are valid.

> On the one hand my suggestion is probably not very sound, on the other hand it worked for me and it is a troubleshooting section after all. You tell me

hard too say. too many moving parts.

i still think we can remove this last paragraph and based on my understanding of the kubelet <-> apiserver setup.
full CP component cert rotation and restart in HA setups is a complicated and we have to have good evidence about it.